go/vt/vttablet/tabletmanager/rpc_replication.go

```diff
@@ -642,30 +662,81 @@ func (tm *TabletManager) SlaveWasRestarted(ctx context.Context, parent *topodata
 
 // StopReplicationAndGetStatus stops MySQL replication, and returns the
 // current status.
-func (tm *TabletManager) StopReplicationAndGetStatus(ctx context.Context) (*replicationdatapb.Status, error) {
+func (tm *TabletManager) StopReplicationAndGetStatus(ctx context.Context, stopIOThreadOnly bool) (StopReplicationAndGetStatusResponse, error) {
 	if err := tm.lock(ctx); err != nil {
-		return nil, err
+		return StopReplicationAndGetStatusResponse{}, err
 	}
 	defer tm.unlock()
 
-	// get the status before we stop replication
+	// Get the status before we stop replication.
+	// We need to do this first to bail out if the replica has already been stopped.
 	rs, err := tm.MysqlDaemon.SlaveStatus()
 	if err != nil {
-		return nil, vterrors.Wrap(err, "before status failed")
+		return StopReplicationAndGetStatusResponse{}, vterrors.Wrap(err, "before status failed")
 	}
-	if !rs.SlaveIORunning && !rs.SlaveSQLRunning {
-		// no replication is running, just return what we got
-		return mysql.SlaveStatusToProto(rs), nil
-	}
-	if err := tm.stopSlaveLocked(ctx); err != nil {
-		return nil, vterrors.Wrap(err, "stop slave failed")
+	before := mysql.SlaveStatusToProto(rs)
+
+	if stopIOThreadOnly {
+		if !rs.SlaveIORunning {
+			return StopReplicationAndGetStatusResponse{
+				Status: before,
+				Before: before,
+			}, nil
+		}
+		if err := tm.stopSlaveIOThreadLocked(ctx); err != nil {
+			return StopReplicationAndGetStatusResponse{
+				Before: before,
+			}, vterrors.Wrap(err, "stop slave io thread failed")
+		}
+	} else {
+		if !rs.SlaveIORunning && !rs.SlaveSQLRunning {
```

Comment:
There is a flaw of logic here. `SlaveIORunning` is read from `SHOW SLAVE STATUS` by checking if the value `== "Yes"`. 

But `Yes` and `No` are not the only two options. The thread may be `Connecting`. In which case, it's not _working_ per se, but actively _trying to_. 

The above logic diagnoses the state as "IO thread not running therefore do _this_", where the truth is the `Connecting` state can turn into a `Yes` any time; the thread must be stopped. So, the IO thread is really not running if it's a `"No"`. But, then, does it even exist?

Again, this is something `orchestrator` has seen some time ago. Consider:

https://github.com/openark/orchestrator/blob/761be91299ce81abf6c9a3889a660549b8daef10/go/inst/instance.go#L60-L61

and

https://github.com/openark/orchestrator/blob/761be91299ce81abf6c9a3889a660549b8daef10/go/inst/replication_thread_state.go#L21-L39

to get some sense of how this is implemented in `orchestrator`.

Reply:
It sounds to me then that we should only ever return early if both are explicitely "No" rather than `not "Yes"`.

Reply:
Because "Connecting" could turn into a yes, so it still represents a case where we should try to issue `stop slave`. Am I correct in this thinking?

Reply:
> It sounds to me then that we should only ever return early if both are explicitely "No" rather than not "Yes".

I think it's best if the status is not a boolean. What if the thread doesn't exist? Does that means it's `!= "No"` thereforce presumed running?

Reply:
> Because "Connecting" could turn into a yes, so it still represents a case where we should try to issue stop slave. Am I correct in this thinking?

Precisely so.

Reply:
For backwards compatibility reasons, we can't change the status type to be non-boolean. We can add a new field to represent other states, like `SlaveIOConnecting` for instance.

Reply:
@enisoc @deepthi thoughts here on the best way to handle representing the "Connecting" state?

Reply:
@PrismaPhonic and I looked at how this property is used in the rest of the code. @shlomi-noach is absolutely correct, that if the state is "Connecting", there are places where we should be stopping replication, and we currently don't do that. Changing the definition of `SlaveIORunning` from `== 'Yes'` to `!= 'No'` sounds right. Alternatively, we could explicitly check for `'Yes' || 'Connecting'`. We also looked at the mysql source, and it's quite clear that these are the only 3 possible values.

Reply:
Perhaps we could add a pair of new fields to our status struct (and the corresponding proto message) called `IOThreadState` and `SQLThreadState` which are enums?

Reply:
Or could we just always stop without checking first? I know you tried that and there was a test failure, but did we determine with certainty that the test failure was caused by "stop" not being idempotent?

Reply:
I believe that would be confusing because there already is a `Slave_IO_State` field that returns from `show slave status`, which is a very different thing. Also, as far as we know this is only an issue with IO and not sql thread. I can double-check on whether that was what exactly caused the failure and try that approach again (stop no matter what).

Regardless, this should be fixed. Deepthi and I found other uses of `SlaveIORunning` where the value determines whether we should call `stopSlave` (such as in `SetMaster`). In reality, we should call stop slave anytime it's not "NO", rather than anything it's "Yes". Fixing this logic will correct other misuses of the field in our codebase.

Reply:
> anytime it's not "NO"

And also not empty

Reply:
@deepthi and I looked at the MySQL codebase and empty should not be possible. Have you encountered empty?

Reply:
Sorry, I was unclear. "Empty" as in "there is no IO thread"

Reply:
(ie the server is not a replica)

Reply:
I think we already handle that case if I'm not mistaken, and return `ErrNotSlave`